dipy/denoise/fast_lpca.pyx

```diff
@@ -0,0 +1,286 @@
+
+                    # Add displaced signed corners
+                    if i >= m:
+                        x = I[i - m, j, k]  # q=(1, 0, 0)
+                        update_outer_prod(cov, x, -1)
+                        update_vector(mu, x, -1)
+                        if j >= m:
+                            x = I[i - m, j - m, k]  # q=(1, 1, 0)
+                            update_outer_prod(cov, x, 1)
+                            update_vector(mu, x, 1)
+                            if k >= m:
+                                x = I[i - m, j - m, k - m]  # q=(1, 1, 1)
+                                update_outer_prod(cov, x, -1)
+                                update_vector(mu, x, -1)
+                        if k >= m:
+                            x = I[i - m, j, k - m]  # q=(1, 0, 1)
+                            update_outer_prod(cov, x, 1)
+                            update_vector(mu, x, 1)
+                    if j >= m:
+                        x = I[i, j - m, k]  # q=(0, 1, 0)
+                        update_outer_prod(cov, x, -1)
+                        update_vector(mu, x, -1)
+                        if k >= m:
+                            x = I[i, j - m, k - m]  # q=(0, 1, 1)
+                            update_outer_prod(cov, x, 1)
+                            update_vector(mu, x, 1)
+                    if k >= m:
+                        x = I[i, j, k - m]  # q=(0, 0, 1)
+                        update_outer_prod(cov, x, -1)
+                        update_vector(mu, x, -1)
+
+                    # Save current integrals for future reference
+                    update_matrix(T[cur_i, j, k], cov, 0)
+                    update_vector(S[cur_i, j, k], mu, 0)
+
+                    # Use integral of current rectangle
+                    # Last corner is (i, j, k) => center is (i-radius,
+                    # j-radius, k-radius)
+                    if (i >= m - 1) and (j >= m - 1) and (k >= m - 1):
+                        # Divide by the number of samples
+                        for ii in range(ndiff):
+                            mu[ii] /= nsamples
+                            for jj in range(ndiff):
+                                cov[ii, jj] /= nsamples
+                        # Subtract self-outer product of mu from cov
+                        update_outer_prod(cov, mu, -1)
+
+                        i0 = i - radius
+                        j0 = j - radius
+                        k0 = k - radius
+
+                        with gil:
+                            # eigen value decomposition
+                            P[...] = 0
+                            d, W_hat = np.linalg.eigh(cov)
+
+                        l0norm = 0
+                        for p in range(ndiff):
+                            if d[p] >= sigma[i0, j0, k0] * 2.3 * 2.3:
+                                l0norm += 1
+                                update_outer_prod(P, W_hat[:, p], 1)
```

Comment:
If we refactor `fast_vecmat_mul` then `P` is now `double **` so we may need a new `update_outer_prod` as well